src/main/flight/autopilot.c

```diff
@@ -22,30 +22,85 @@
 
 #include "platform.h"
 #include "build/debug.h"
+#include "common/filter.h"
 #include "common/maths.h"
+#include "fc/rc.h"
+#include "fc/runtime_config.h"
 
 #include "flight/imu.h"
-#include "flight/pid.h"
 #include "flight/position.h"
 #include "rx/rx.h"
+#include "sensors/gyro.h"
+#include "sensors/compass.h"
 
 #include "autopilot.h"
 
 #define ALTITUDE_P_SCALE  0.01f
 #define ALTITUDE_I_SCALE  0.003f
 #define ALTITUDE_D_SCALE  0.01f
 #define ALTITUDE_F_SCALE  0.01f
+#define POSITION_P_SCALE  0.001f
+#define POSITION_I_SCALE  0.0004f
+#define POSITION_D_SCALE  0.003f
+#define POSITION_A_SCALE  0.0008f
 
 static pidCoefficient_t altitudePidCoeffs;
+static pidCoefficient_t positionPidCoeffs;
+
 static float altitudeI = 0.0f;
 static float throttleOut = 0.0f;
 
+typedef struct {
+    float distanceCm;
+    float previousDistanceCm;
+    float sanityCheckDistance;
+    float previousVelocity;
+    float initialHeadingDeg;
+    float pitchI;
+    float rollI;
+    bool isStarting;
+    float peakInitialGroundspeed;
+    float lpfCutoff;
+    bool sticksActive;
+} posHoldState;
+
+static posHoldState posHold = {
+    .distanceCm = 0.0f,
+    .previousDistanceCm = 0.0f,
+    .sanityCheckDistance = 1000.0f,
+    .previousVelocity = 0.0f,
+    .initialHeadingDeg = 0.0f,
+    .pitchI = 0.0f,
+    .rollI = 0.0f,
+    .isStarting = false,
+    .peakInitialGroundspeed = 0.0f,
+    .lpfCutoff = 1.0f,
+    .sticksActive = false,
+};
+
+static gpsLocation_t currentTargetLocation = {0, 0, 0};
+float autopilotAngle[ANGLE_INDEX_COUNT];
+static pt1Filter_t velocityPitchLpf;
+static pt1Filter_t velocityRollLpf;
+static pt2Filter_t accelerationRollLpf;
+static pt2Filter_t accelerationPitchLpf;
+
 void autopilotInit(const autopilotConfig_t *config)
 {
     altitudePidCoeffs.Kp = config->altitude_P * ALTITUDE_P_SCALE;
     altitudePidCoeffs.Ki = config->altitude_I * ALTITUDE_I_SCALE;
     altitudePidCoeffs.Kd = config->altitude_D * ALTITUDE_D_SCALE;
     altitudePidCoeffs.Kf = config->altitude_F * ALTITUDE_F_SCALE;
+    positionPidCoeffs.Kp = config->position_P * POSITION_P_SCALE;
+    positionPidCoeffs.Ki = config->position_I * POSITION_I_SCALE;
+    positionPidCoeffs.Kd = config->position_D * POSITION_D_SCALE;
+    positionPidCoeffs.Kf = config->position_A * POSITION_A_SCALE; // Kf used for acceleration
+    posHold.lpfCutoff = config->position_cutoff * 0.01f;
+    const float pt1Gain = pt1FilterGain(posHold.lpfCutoff, 0.1f); // assume 10Hz GPS connection at start
+    pt1FilterInit(&velocityRollLpf, pt1Gain);
+    pt1FilterInit(&velocityPitchLpf, pt1Gain);
+    pt2FilterInit(&accelerationRollLpf, pt1Gain);
+    pt2FilterInit(&accelerationPitchLpf, pt1Gain);
```

Comment:
"Faulty" filter gain

Reply:
Deliberate, to re-use gain parameter and provide stronger filtering for PT2.  Previous testing used series PT1 at same cutoff, and that works well, so I just did this to get the same result.

Reply:
Then this might be worth a comment. Otherwise someone might accidentally "correct" it because it looks like a bug.
Maybe something like this in each line that recycles the pt1Gain:
```c
// uses pt1Gain on purpose (PT2 cutoff will be ~36% lower than PT1 cutoff)
```

Reply:
added